### PR TITLE
Add more details to architecture introduction

### PR DESCRIPTION
**Pull request by tfpauly**
Address issue #189 
- Add more intro section, with friendly diagrams
- Highlight why we use events, messages, and racing

### CODE REVIEW

File: draft-ietf-taps-arch.md [hunk 1]
```diff
@@ -115,6 +115,82 @@ The goal of the Transport Services architecture is to provide a common, flexible
...
+                          |
++-----------------------------------------------------+
+|            Kernel Protocol Implementation           |
+|                    (UDP, TCP)                       |
++-----------------------------------------------------+
+                          |
++-----------------------------------------------------+
+|               Network Layer Interface               |
++-----------------------------------------------------+
+
```
Comment (mwelzl) [inline draft-ietf-taps-arch.md]: I like this, but would propose to update it as follows:

```
 +-----------------------------------------------------+
 |                    Application                      |
 +-----------------------------------------------------+
        |                  |                    |
 +-----------------------------------------------------+
 |      |             Socket API                |      |
 +-----------------------------------------------------+
        |                                       |
 +-----------------------------------------------------+
 |      |     Kernel Protocol Implementation    |      |
 |     TCP                                     UDP     |
 +-----------------------------------------------------+
                           |
 +-----------------------------------------------------+
 |               Network Layer Interface               |
 +-----------------------------------------------------+
```

File: draft-ietf-taps-arch.md [hunk 2]
```diff
@@ -115,6 +115,82 @@ The goal of the Transport Services architecture is to provide a common, flexible
...
+All interaction with a Transport Services system is expected to be asynchronous, and use an event-driven model unlike sockets {{events}}. For example, if the application wants to read, its call to read will not fail, but will deliver an event containing the received data once it is available.
+
+The Transport Services API also delivers events regarding the lifetime of a connection and changes to available network links, which were not previously made explicit in sockets.
+
+Using asynchronous events allows for a much simpler interaction model when establishing connections and transferring data. Events in time more closely reflects the nature of interactions over networks, as opposed to how sockets represent network resources as file system objects that may be temporarily unavailable.
```
Comment (mwelzl) [inline draft-ietf-taps-arch.md]: A nit - should this be: "Events in time more closely *reflect* "  ?

File: draft-ietf-taps-arch.md (hunk 1 above)
Comment (tfpauly) [reply draft-ietf-taps-arch.md]: What are the lines from the application down to TCP and UDP? Just showing the different use cases?

Comment (mwelzl) [reply draft-ietf-taps-arch.md]: No, the point was to show that applications are statically bound to either TCP or UDP.

Comment (tfpauly) [reply draft-ietf-taps-arch.md]: Ah, I see now. That makes more sense now. I'll try to redraw to illustrate that.

Comment (tfpauly) [reply draft-ietf-taps-arch.md]: Okay, take a look at the updated diagram

Comment (mwelzl) [reply draft-ietf-taps-arch.md]: Much nicer! Thanks for this!

File: draft-ietf-taps-arch.md [hunk 3]
```diff
@@ -115,6 +115,82 @@ The goal of the Transport Services architecture is to provide a common, flexible
...
+|                    Application                      |
++-----------------------------------------------------+
+                          |
++-----------------------------------------------------+
+|               Transport Services API                |
++-----------------------------------------------------+
+                          |
++-----------------------------------------------------+
+|           Transport System Implementation           |
+|          (UDP, TCP, SCTP, DCCP, QUIC, etc)          |
```
[the remark this answers is not in the document]
Comment (tfpauly) [reply draft-ietf-taps-arch.md]: Yes we should!

File: draft-ietf-taps-arch.md [hunk 4]
```diff
@@ -115,6 +115,82 @@ The goal of the Transport Services architecture is to provide a common, flexible
...
+The model for using sockets for networking can be represented as follows: applications create connections and transfer data using the socket API, which provides the interface to the implementations of UDP and TCP (typically implemented in the system's kernel), which in turn send data over the available network layer interfaces.
+
+~~~~~~~~~~
+
++-----------------------------------------------------+
+|                    Application                      |
++-----------------------------------------------------+
+           |                             |
++---------------------+       +-----------------------+
+|  Socket Stream API  |       |  Socket Datagram API  |
```
[the remark this answers is not in the document]
Comment (tfpauly) [reply draft-ietf-taps-arch.md]: I think it's a bit messy for the diagram. I've updated more of the text below to mention multiple IPs and higher level endpoints below.

[the remark this answers is not in the document]
Comment (chris-wood) [reply draft-ietf-taps-arch.md]: 👍

**Pull request merged**